### PR TITLE
Blacklist default globals to support exotic user code with `__` prefixed annotations

### PR DESCRIPTION
**Pull request by sydney-runkle**
This way, if someone uses something like `__test__` as a type annotation, we don't remove that from the namespace.

Contributing to https://github.com/pydantic/pydantic/issues/10074

### CODE REVIEW

File: pydantic/_internal/_typing_extra.py [hunk 1]
```diff
@@ -167,9 +167,22 @@ def is_finalvar(ann_type: Any) -> bool:
...
+    '__builtins__',
+    '__file__',
+    '__cached__',
+]
+
+
 def _remove_default_globals_from_ns(namespace: dict[str, Any]) -> dict[str, Any]:
     """Remove default globals like __name__, __doc__, etc that aren't needed for type evaluation."""
-    return {k: v for k, v in namespace.items() if not k.startswith(('__', '@'))}
+    return {k: v for k, v in namespace.items() if not k.startswith('@') and k not in _DEFAULT_GLOBALS}
```
Comment (Viicos) [inline pydantic/_internal/_typing_extra.py:185]: I'm wondering, `@` was taken from pytest-examples but is it really possible to have something starting with that in the namespace? `@.*` isn't a valid identifier

**Review by Viicos — APPROVED** [review]

File: pydantic/_internal/_typing_extra.py (hunk 1 above)
Comment (sydney-runkle) [reply pydantic/_internal/_typing_extra.py:185]: Good question, yeah, I saw lots of vars starting with `@` in the module namespace when I was running `pytest`, but perhaps that's not important for real applications. I guess I'm inclined to keep it bc it seems like the performance drawback is negligible, and it ensures the namespace is pretty clean.

**Comment by sydney-runkle** [comment]
Actually, I'd like to wait on merging this, I'm seeing this slow down the `get_cls_types_namespace` function significantly. Perhaps I'll remove the `@`, or at least consider the effects.

**Comment by sydney-runkle** [comment]
Alright, removing the double check made it such that we're not seeing any significant performance changes. Going to merge :).

**Pull request merged**